### ISSUE TITLE
test failures on Power8 machine

### ISSUE THREAD

**Issue by sharkcz** [openssl/openssl#19163]
I am seeing test failures when running the test suite from the master branch on Power8 machines (both bare metal and VM) running Fedora/ppc64le. The test report looks like
````
Test Summary Report
-------------------
30-test_evp.t                    (Wstat: 256 (exited 1) Tests: 74 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
70-test_asyncio.t                (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
70-test_comp.t                   (Wstat: 27392 (exited 107) Tests: 0 Failed: 0)
  Non-zero exit status: 107
  Parse errors: No plan found in TAP output
70-test_key_share.t              (Wstat: 27392 (exited 107) Tests: 0 Failed: 0)
  Non-zero exit status: 107
  Parse errors: No plan found in TAP output
70-test_recordlen.t              (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
70-test_servername.t             (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
70-test_sslextension.t           (Wstat: 27392 (exited 107) Tests: 7 Failed: 0)
  Non-zero exit status: 107
  Parse errors: Bad plan.  You planned 8 tests but ran 7.
70-test_sslrecords.t             (Wstat: 27392 (exited 107) Tests: 12 Failed: 0)
  Non-zero exit status: 107
  Parse errors: Bad plan.  You planned 21 tests but ran 12.
70-test_sslsigalgs.t             (Wstat: 27392 (exited 107) Tests: 0 Failed: 0)
  Non-zero exit status: 107
  Parse errors: No plan found in TAP output
70-test_sslsignature.t           (Wstat: 27392 (exited 107) Tests: 0 Failed: 0)
  Non-zero exit status: 107
  Parse errors: No plan found in TAP output
70-test_sslversions.t            (Wstat: 27392 (exited 107) Tests: 4 Failed: 0)
  Non-zero exit status: 107
  Parse errors: Bad plan.  You planned 8 tests but ran 4.
70-test_tls13alerts.t            (Wstat: 27392 (exited 107) Tests: 0 Failed: 0)
  Non-zero exit status: 107
  Parse errors: No plan found in TAP output
70-test_tls13cookie.t            (Wstat: 27392 (exited 107) Tests: 0 Failed: 0)
  Non-zero exit status: 107
  Parse errors: No plan found in TAP output
70-test_tls13downgrade.t         (Wstat: 27392 (exited 107) Tests: 4 Failed: 0)
  Non-zero exit status: 107
  Parse errors: Bad plan.  You planned 6 tests but ran 4.
70-test_tls13hrr.t               (Wstat: 27392 (exited 107) Tests: 0 Failed: 0)
  Non-zero exit status: 107
  Parse errors: No plan found in TAP output
70-test_tls13kexmodes.t          (Wstat: 27392 (exited 107) Tests: 0 Failed: 0)
  Non-zero exit status: 107
  Parse errors: No plan found in TAP output
70-test_tls13messages.t          (Wstat: 28416 (exited 111) Tests: 0 Failed: 0)
  Non-zero exit status: 111
  Parse errors: No plan found in TAP output
70-test_tls13psk.t               (Wstat: 27392 (exited 107) Tests: 0 Failed: 0)
  Non-zero exit status: 107
  Parse errors: No plan found in TAP output
70-test_tlsextms.t               (Wstat: 27392 (exited 107) Tests: 9 Failed: 0)
  Non-zero exit status: 107
  Parse errors: Bad plan.  You planned 10 tests but ran 9.
80-test_dtls_mtu.t               (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
80-test_ssl_new.t                (Wstat: 6144 (exited 24) Tests: 31 Failed: 24)
  Failed tests:  1-18, 20-21, 24, 26-28
  Non-zero exit status: 24
80-test_ssl_old.t                (Wstat: 512 (exited 2) Tests: 7 Failed: 2)
  Failed tests:  2-3
  Non-zero exit status: 2
80-test_sslcorrupt.t             (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
90-test_sslapi.t                 (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
90-test_sslbuffers.t             (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
90-test_tls13ccs.t               (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
90-test_tls13encryption.t        (Wstat: 256 (exited 1) Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
99-test_fuzz_client.t            (Wstat: 256 (exited 1) Tests: 2 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
99-test_fuzz_server.t            (Wstat: 256 (exited 1) Tests: 2 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
Files=256, Tests=3316, 562 wallclock secs (10.93 usr  0.69 sys + 495.51 cusr 36.15 csys = 543.28 CPU)
Result: FAIL
````
The symptoms are SIGILL dumps/tracebacks on the console with corresponding entries in the journal
````
[15978.695058] evp_test[65741]: illegal instruction (4) at 7fffafa1c9ec nip 7fffafa1c9ec lr 7fffafb03c74 code 1 in libcrypto.so.3[7fffaf840000+350000]
[15978.695247] evp_test[65741]: code: 7f4909ce 39290010 7f6909ce 39290010 7f8909ce 39290010 7fa909ce 39290010 
[15978.695264] evp_test[65741]: code: 7fc909ce 39290010 7fe909ce f8010210 <7c0046d9> 39400020 7c4a4699 39400030 
[16002.918406] asynciotest[66770]: illegal instruction (4) at 7fff8082c9ec nip 7fff8082c9ec lr 7fff80913c74 code 1 in libcrypto.so.3[7fff80650000+350000]
[16002.918564] asynciotest[66770]: code: 7f4909ce 39290010 7f6909ce 39290010 7f8909ce 39290010 7fa909ce 39290010 
[16002.918585] asynciotest[66770]: code: 7fc909ce 39290010 7fe909ce f8010210 <7c0046d9> 39400020 7c4a4699 39400030 
...
````
The suspicious changes are commit 44a563dde1 (and/or commit 345c99b6654b).

Please see also https://bugzilla.redhat.com/show_bug.cgi?id=2124845 for more details and history.

**Comment by t8m** [comment]
@dannytsen Could you please take a look?

**Comment by neverpanic** [comment]
It seems the choice to define `AES_gcm_encrypt` and `AES_gcm_decrypt` in https://github.com/openssl/openssl/commit/44a563dde1584cd9284e80b6e45ee5019be8d36c#diff-c71814c0da77a0992352841da95a677f31d461cc6b9f63eba96a6fab9faaa2b0R85-R86 caused this code to be unconditionally enabled for all PPC CPUs without a runtime check. This should probably be reverted.

**Comment by sharkcz** [comment]
this is the backtrace from one of the tests
````
(gdb) where
#0  ppc_aes_gcm_encrypt () at crypto/modes/aes-gcm-ppc.s:419
#1  0x00007fffbd225310 in aes_p10_gcm_crypt (in=in@entry=0x1003a017ea8 "\v", out=out@entry=0x1003a017ea8 "\v", len=368, len@entry=369, key=0x1003a03a2d8, 
    ivec=ivec@entry=0x1003a03a108 "xݞ\210\322\321Ӂ\343suh", Xi=Xi@entry=0x1003a03a148, encrypt=encrypt@entry=1) at providers/implementations/ciphers/cipher_aes_gcm_hw_ppc.inc:76
#2  0x00007fffbd22adf8 in ppc_aes_gcm_encrypt_wrap (Xi=0x1003a03a148, ivec=0x1003a03a108 "xݞ\210\322\321Ӂ\343suh", key=<optimized out>, len=369, out=0x1003a017ea8 "\v", 
    in=0x1003a017ea8 "\v") at providers/implementations/ciphers/cipher_aes_gcm_hw_ppc.inc:96
#3  generic_aes_gcm_cipher_update (ctx=0x1003a03a010, in=0x1003a017ea8 "\v", len=369, out=0x1003a017ea8 "\v") at providers/implementations/ciphers/cipher_aes_gcm_hw.c:76
#4  0x00007fffbd290b20 in gcm_cipher_internal (ctx=0x1003a03a010, out=0x1003a017ea8 "\v", padlen=0x7fffdb03b710, in=0x1003a017ea8 "\v", len=369)
    at providers/implementations/ciphers/ciphercommon_gcm.c:426
#5  0x00007fffbd290e50 in ossl_gcm_stream_update (vctx=<optimized out>, out=<optimized out>, outl=<optimized out>, outsize=<optimized out>, in=<optimized out>, inl=<optimized out>)
    at providers/implementations/ciphers/ciphercommon_gcm.c:319
#6  0x00007fffbd10b130 in EVP_EncryptUpdate (ctx=<optimized out>, out=<optimized out>, outl=0x7fffdb03b7b4, in=<optimized out>, inl=<optimized out>) at crypto/evp/evp_enc.c:649
#7  0x00007fffbd10be84 in EVP_CipherUpdate (ctx=<optimized out>, out=<optimized out>, outl=<optimized out>, in=<optimized out>, inl=<optimized out>) at crypto/evp/evp_enc.c:425
#8  0x00007fffbd4adfc8 in tls13_enc (s=0x1003a037120, recs=0x7fffdb03c018, n_recs=<optimized out>, sending=<optimized out>, mac=<optimized out>, macsize=<optimized out>)
    at ssl/record/ssl3_record_tls13.c:158
#9  0x00007fffbd4b6044 in do_ssl3_write (s=s@entry=0x1003a037120, type=type@entry=22, buf=buf@entry=0x1003a012940 "\v", pipelens=pipelens@entry=0x7fffdb03ca58, numpipes=1, 
    create_empty_fragment=create_empty_fragment@entry=0, written=written@entry=0x7fffdb03ca50) at ssl/record/rec_layer_s3.c:873
#10 0x00007fffbd4b6908 in ssl3_write_bytes (ssl=0x1003a037120, type=<optimized out>, buf_=0x1003a012940, len=<optimized out>, written=0x7fffdb03cbe0) at ssl/record/rec_layer_s3.c:460
#11 0x00007fffbd4dadc8 in ssl3_do_write (s=<optimized out>, type=<optimized out>) at ssl/statem/statem_lib.c:49
#12 0x00007fffbd475bfc in ssl3_handshake_write (s=<optimized out>) at ssl/s3_lib.c:3327
#13 0x00007fffbd4d1ad4 in statem_do_write (s=0x1003a037120) at ssl/statem/statem.c:756
#14 write_state_machine (s=0x1003a037120) at ssl/statem/statem.c:909
#15 state_machine (s=0x1003a037120, server=<optimized out>) at ssl/statem/statem.c:487
#16 0x00007fffbd496ae4 in SSL_do_handshake (s=0x1003a037120) at ssl/ssl_lib.c:4430
#17 0x0000000129121564 in FuzzerTestOneInput (buf=buf@entry=0x10039fbe6c0 "\026\003\001\002]\001", len=610) at fuzz/server.c:640
#18 0x0000000129121888 in FuzzerTestOneInput (len=<optimized out>, buf=0x10039fbe6c0 "\026\003\001\002]\001") at fuzz/server.c:538
#19 testfile (pathname=pathname@entry=0x10039fb9760 "../../fuzz/corpora/server/05f335262f1bb688536e8a65ed0745243f0be385") at fuzz/test-corpus.c:55
#20 0x0000000129120a08 in main (argc=<optimized out>, argv=<optimized out>) at fuzz/test-corpus.c:90
````

**Comment by sharkcz** [comment]
````
diff --git a/include/crypto/aes_platform.h b/include/crypto/aes_platform.h
index 72f217ce4f..7eabba4362 100644
--- a/include/crypto/aes_platform.h
+++ b/include/crypto/aes_platform.h
@@ -89,8 +89,8 @@ size_t ppc_aes_gcm_encrypt_wrap(const unsigned char *in, unsigned char *out,
 size_t ppc_aes_gcm_decrypt_wrap(const unsigned char *in, unsigned char *out,
                                 size_t len, const void *key,
                                 unsigned char ivec[16], u64 *Xi);
-#   define AES_gcm_encrypt ppc_aes_gcm_encrypt_wrap
-#   define AES_gcm_decrypt ppc_aes_gcm_decrypt_wrap
+#   define AES_gcm_encrypt ppc_aes_gcm_encrypt
+#   define AES_gcm_decrypt ppc_aes_gcm_decrypt
 #   define AES_GCM_ASM(gctx) ((gctx)->ctr==aes_p8_ctr32_encrypt_blocks && \
                               (gctx)->gcm.funcs.ghash==gcm_ghash_p8)
 void gcm_ghash_p8(u64 Xi[2],const u128 Htable[16],const u8 *inp, size_t len);
````
might be the fix, checking further ...

**Comment by t8m** [comment]
Could you try to set a breakpoint at `ossl_prov_aes_hw_gcm`? I'm curious how it could select the incorrect implementation on the P8 platform.

**Comment by neverpanic** [comment]
The check in `ossl_prov_aes_hw_gcm` works correctly. The `ppc_aes_gcm_encrypt_wrap` implementation is chosen unconditionally on all PPC architectures in `providers/implementations/ciphers/cipher_aes_gcm_hw.c:generic_aes_gcm_cipher_update` in line 76, because that file includes `include/crypto/aes_platform.h`, which `#define`s `AES_gcm_encrypt` to `ppc_aes_gcm_encrypt_wrap` on all PPC arches.

**Comment by t8m** [comment]
Can you please try this?
```diff
diff --git a/providers/implementations/ciphers/cipher_aes_gcm_hw_ppc.inc b/providers/implementations/ciphers/cipher_aes_gcm_hw_ppc.inc
index 5a1c3e4a58..b9d776f1cc 100644
--- a/providers/implementations/ciphers/cipher_aes_gcm_hw_ppc.inc
+++ b/providers/implementations/ciphers/cipher_aes_gcm_hw_ppc.inc
@@ -90,16 +90,21 @@ static size_t aes_p10_gcm_crypt(const unsigned char *in, unsigned char *out, siz
     return ndone;
 }
 
+#undef AES_gcm_encrypt
+#undef AES_gcm_decrypt
+
 size_t ppc_aes_gcm_encrypt_wrap(const unsigned char *in, unsigned char *out, size_t len,
                                 const void *key, unsigned char ivec[16], u64 *Xi)
 {
-    return aes_p10_gcm_crypt(in, out, len, key, ivec, Xi, 1);
+    return PPC_AES_GCM_CAPABLE ? aes_p10_gcm_crypt(in, out, len, key, ivec, Xi, 1)
+                               : AES_gcm_encrypt(in, out, len, key, ivec, Xi);
 }
 
 size_t ppc_aes_gcm_decrypt_wrap(const unsigned char *in, unsigned char *out, size_t len,
                                 const void *key, unsigned char ivec[16], u64 *Xi)
 {
-    return aes_p10_gcm_crypt(in, out, len, key, ivec, Xi, 0);
+    return PPC_AES_GCM_CAPABLE ? aes_p10_gcm_crypt(in, out, len, key, ivec, Xi, 0)
+                               : AES_gcm_decrypt(in, out, len, key, ivec, Xi);
 }
```

**Comment by sharkcz** [comment]
build fails with
````
/usr/bin/ld: libcrypto.a(libdefault-lib-cipher_aes_gcm_hw.o): in function `ppc_aes_gcm_encrypt_wrap':
/home/sharkcz/projects/openssl/providers/implementations/ciphers/cipher_aes_gcm_hw_ppc.inc:100: undefined reference to `AES_gcm_encrypt'
/usr/bin/ld: libcrypto.a(libdefault-lib-cipher_aes_gcm_hw.o): in function `ppc_aes_gcm_decrypt_wrap':
/home/sharkcz/projects/openssl/providers/implementations/ciphers/cipher_aes_gcm_hw_ppc.inc:107: undefined reference to `AES_gcm_decrypt'
````
but the approach looks correct to me

**Comment by sharkcz** [comment]
it looks good, thanks
````
...
Files=256, Tests=3428, 689 wallclock secs (10.22 usr  1.99 sys + 528.09 cusr 145.24 csys = 685.54 CPU)
Result: PASS
````

**Comment by sharkcz** [comment]
the fix in PR #19182 looks good
